### PR TITLE
ARROW-16633: [C++] Fix incorrect Decimal literal conversion in `ir_consumer.cc`

### PR DESCRIPTION
**Pull request by rz-vastdata**

### CODE REVIEW

**Comment by lidavidm** [comment]
Thanks for the contribution!

Just a couple general points, but first: IIRC we are favoring Substrait over the Compute IR now, should we instead drop this module? CC @westonpace @bkietz 

We should add tests & create a JIRA for this as it's not really a minor PR (even if the fix is only a few lines of code)

**Comment by bkietz** [comment]
We should definitely consider dropping the `!substrait` IR; initially we thought it'd be a useful sibling project for substrait but I no longer believe that's a goal. That said, this is a useful change and I agree that it's not minor; @rzeyde-vastdata if you create a JIRA account we can assign https://issues.apache.org/jira/browse/ARROW-16633 to you

**Comment by rzeyde-vastdata** [comment]
Many thanks!
I've created a Jira user: https://issues.apache.org/jira/secure/ViewProfile.jspa?name=rzeyde-vastdata

@lidavidm What should be the next step(s)?

**Comment by lidavidm** [comment]
Either we should add a test case for this, if we're keeping the compute IR; or we need to decide that we're just going to remove all this code entirely and not bother fixing it.

**Comment by bkietz** [comment]
@rzeyde-vastdata please add a unit test which checks a roundtripped decimal literal

File: cpp/src/arrow/compute/exec/ir_consumer.cc [hunk 1]
```diff
@@ -152,14 +152,14 @@ struct ConvertLiteralImpl {
     switch (t.id()) {
       case Type::DECIMAL128: {
         std::array<uint64_t, 2> little_endian;
-        std::memcpy(little_endian.data(), lit->value(), lit->value()->size());
+        std::memcpy(little_endian.data(), lit->value()->data(), lit->value()->size());
```
Comment (pitrou) [inline cpp/src/arrow/compute/exec/ir_consumer.cc:155]: This still seems incorrect. `little_endian` is default-initialized but, as the `std::array` docs [note](https://en.cppreference.com/w/cpp/container/array), 
> default initialization may result in indeterminate values for non-class T

... meaning that unless `lit->value->size()` is large enough, some decimal data may still be incorrect.

@bkietz Am I wrong?

Comment (bkietz) [reply cpp/src/arrow/compute/exec/ir_consumer.cc:155]: The size of the literal string is asserted to be equal to the decimal's byte_width above, so every byte of `little_endian` should be safely overwritten

Comment (pitrou) [reply cpp/src/arrow/compute/exec/ir_consumer.cc:155]: Woops, I missed that, sorry.

**Comment by pitrou** [comment]
@rzeyde-vastdata Would you like to add a unit test as suggested?

**Comment by rzeyde-vastdata** [comment]
Sorry for the delayed response... unfortunately, I won't be able to add the unit test in upcoming weeks.

**Pull request closed without merging**